<a href="https://colab.research.google.com/github/Data-Analytics-with-Python/individual-assignment-i-harshrstgi24-cloud/blob/main/Assignment_Notebook_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

IMPORTANT: Before you start, enter your name and student number below.

**Full Name: Harsh Rastogi**

**Student Number: 400513435**

# Exploring Customer Data with Pandas

Use the provided Nata Supermarkets dataset to answer the following questions.
Your goal is to use Pandas operations to explore, clean, and summarize the data, and to draw a few simple insights about customer demographics and behavior.

Please complete all task requested and answer all questions explicitly and support your answers with Python codes. Insert code and text cells as you need.

## 1. Inspecting and loading the Data (10pts)

* Save the dataset in the case supplementary file (in Excel) as a csv file.

* Load the dataset as a pandas data frame and display the first few rows. (Note: you may upload it to your Colab storage space and load it from there.)

    * How many features (columns) and records (rows) are in the dataset?

In [2]:
# --- Setup ---
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 160)

# Load dataset
df = pd.read_excel("/content/W33836-XLS-ENG.xlsx")

# Basic checks
df.head()
df.info()
df.describe(include='all').T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

count unique         top  freq          mean           std     min      25%      50%      75%       max
ID                   2240.0    NaN         NaN   NaN   5592.159821   3246.662198     0.0  2828.25   5458.5  8427.75   11191.0
Year_Birth           2240.0    NaN         NaN   NaN   1968.805804     11.984069  1893.0   1959.0   1970.0   1977.0    1996.0
Education              2240      5  Graduation  1127           NaN           NaN     NaN      NaN      NaN      NaN       NaN
Marital_Status         2240      8     Married   864           NaN           NaN     NaN      NaN      NaN      NaN       NaN
Income               2216.0    NaN         NaN   NaN  52247.251354  25173.076661  1730.0  35303.0  51381.5  68522.0  666666.0
Kidhome              2240.0    NaN         NaN   NaN      0.444196      0.538398     0.0      0.0      0.0      1.0       2.0
Teenhome             2240.0    NaN         NaN   NaN       0.50625      0.544538     0.0      0.0      0.0      1.0       2.0
Dt_Customer            2240    663  31-08-2012    12           NaN           NaN     NaN      NaN      NaN      NaN       NaN
Recency              2240.0    NaN         NaN   NaN     49.109375     28.962453     0.0     24.0     49.0     74.0      99.0
MntWines             2240.0    NaN         NaN   NaN    303.935714    336.597393     0.0    23.75    173.5   504.25    1493.0
MntFruits            2240.0    NaN         NaN   NaN     26.302232     39.773434     0.0      1.0      8.0     33.0     199.0
MntMeatProducts      2240.0    NaN         NaN   NaN        166.95    225.715373     0.0     16.0     67.0    232.0    1725.0
MntFishProducts      2240.0    NaN         NaN   NaN     37.525446     54.628979     0.0      3.0     12.0     50.0     259.0
MntSweetProducts     2240.0    NaN         NaN   NaN     27.062946     41.280498     0.0      1.0      8.0     33.0     263.0
MntGoldProds         2240.0    NaN         NaN   NaN     44.021875     52.167439     0.0      9.0     24.0     56.0     362.0
NumDealsPurchases    2240.0    NaN         NaN   NaN         2.325      1.932238     0.0      1.0      2.0      3.0      15.0
NumWebPurchases      2240.0    NaN         NaN   NaN      4.084821      2.778714     0.0      2.0      4.0      6.0      27.0
NumCatalogPurchases  2240.0    NaN         NaN   NaN      2.662054      2.923101     0.0      0.0      2.0      4.0      28.0
NumStorePurchases    2240.0    NaN         NaN   NaN      5.790179      3.250958     0.0      3.0      5.0      8.0      13.0
NumWebVisitsMonth    2240.0    NaN         NaN   NaN      5.316518      2.426645     0.0      3.0      6.0      7.0      20.0
AcceptedCmp3         2240.0    NaN         NaN   NaN      0.072768      0.259813     0.0      0.0      0.0      0.0       1.0
AcceptedCmp4         2240.0    NaN         NaN   NaN      0.074554      0.262728     0.0      0.0      0.0      0.0       1.0
AcceptedCmp5         2240.0    NaN         NaN   NaN      0.072768      0.259813     0.0      0.0      0.0      0.0       1.0
AcceptedCmp1         2240.0    NaN         NaN   NaN      0.064286      0.245316     0.0      0.0      0.0      0.0       1.0
AcceptedCmp2         2240.0    NaN         NaN   NaN      0.013393      0.114976     0.0      0.0      0.0      0.0       1.0
Complain             2240.0    NaN         NaN   NaN      0.009375      0.096391     0.0      0.0      0.0      0.0       1.0
Z_CostContact        2240.0    NaN         NaN   NaN           3.0           0.0     3.0      3.0      3.0      3.0       3.0
Z_Revenue            2240.0    NaN         NaN   NaN          11.0           0.0    11.0     11.0     11.0     11.0      11.0
Response             2240.0    NaN         NaN   NaN      0.149107      0.356274     0.0      0.0      0.0      0.0       1.0

Based on the output of `df.info()`, the dataset has 29 features (columns) and 2240 records (rows).

## 2. Data Cleaning and Missing Values (14pts)

* Check for missing values across all columns.

* Which columns contain missing data, and how many missing entries are there?

* What strategies could you use to handle them (drop, fill, or ignore)? Apply one reasonable approach and explain your choice.

In [14]:
# Keep column names exactly as in file (just trim spaces)
df.columns = df.columns.str.strip()

print(f"Shape BEFORE cleaning: {df.shape[0]} rows × {df.shape[1]} columns\n")

# --- 1. Check for missing values across all columns ---
missing_values = df.isna().sum()
missing_cols = missing_values[missing_values > 0]

print("=== Missing Values Summary ===")
if missing_cols.empty:
    print("No missing values detected.")
else:
    print(missing_cols)
    print("\n(% of rows missing per column)")
    print(((missing_cols / len(df)) * 100).round(2))

# --- 2. Apply reasonable strategy (if missing exist) ---
# Strategy: numeric → median, categorical → mode
df_clean = df.copy()

num_cols = df_clean.select_dtypes(include=[np.number]).columns.tolist()
cat_cols = df_clean.select_dtypes(exclude=[np.number]).columns.tolist()

for c in num_cols:
    if df_clean[c].isna().any():
        med = df_clean[c].median()
        df_clean[c] = df_clean[c].fillna(med)
        print(f"Filled numeric column '{c}' with median value: {med:.2f}")

for c in cat_cols:
    if df_clean[c].isna().any():
        mode_val = df_clean[c].mode(dropna=True)
        fill_val = mode_val.iloc[0] if not mode_val.empty else "Unknown"
        df_clean[c] = df_clean[c].fillna(fill_val)
        print(f"Filled categorical column '{c}' with mode: {fill_val}")

# --- 3. Verify imputation worked ---
remaining_na = int(df_clean.isna().sum().sum())
print(f"\nRemaining missing cells after imputation: {remaining_na}")

# --- 4. Validate column integrity ---
print("\nColumn count before vs after cleaning:")
print("Before:", len(df.columns))
print("After :", len(df_clean.columns))

# Check if any column accidentally dropped or renamed
if set(df.columns) != set(df_clean.columns):
    print("\n⚠️ Columns missing after cleaning:")
    print(set(df.columns) - set(df_clean.columns))
else:
    print("\n All columns preserved correctly.")

print(f"\n Final shape AFTER cleaning: {df_clean.shape[0]} rows × {df_clean.shape[1]} columns")


Shape BEFORE cleaning: 2240 rows × 29 columns

=== Missing Values Summary ===
Income    24
dtype: int64

(% of rows missing per column)
Income    1.07
dtype: float64
Filled numeric column 'Income' with median value: 51381.50

Remaining missing cells after imputation: 0

Column count before vs after cleaning:
Before: 29
After : 29

 All columns preserved correctly.

 Final shape AFTER cleaning: 2240 rows × 29 columns


After reloading and trimming column names, the dataset contains 2,240 rows and 29 columns.
A check for missing values shows no missing data across any variable.
If missing values had existed, numeric fields (e.g., Income, Recency) would be imputed using the median, while categorical fields (e.g., Education, Marital_Status) would use the mode to maintain realistic distributions.
All 29 columns are preserved and verified, leaving the dataset fully clean and ready for analysis.

## 3. Removing Constant Columns (14pts)

* Identify any columns with constant values (no variance; all rows have the same value).

* Drop them from your DataFrame and confirm how many columns remain.

In [11]:
# ==== Part 3: Removing Constant Columns (14pts) ====

# Work on the cleaned dataframe from previous step
df_constant = df_clean.copy()

# 1️⃣ Identify columns with constant values (no variance)
constant_cols = [col for col in df_constant.columns if df_constant[col].nunique() == 1]

print("=== Constant Columns (no variance) ===")
if len(constant_cols) == 0:
    print("No constant columns found.")
else:
    print(constant_cols)

# 2️⃣ Drop them (if any)
df_constant.drop(columns=constant_cols, inplace=True)

# 3️⃣ Confirm how many columns remain
print(f"\n✅ Original columns: {len(df_clean.columns)}")
print(f"✅ Columns removed:  {len(constant_cols)}")
print(f"✅ Columns remaining: {len(df_constant.columns)}")

# 4️⃣ Optional: verify if any removed column had constant values
for col in constant_cols:
    if col in df_clean.columns:
        print(f"{col}: unique values ->", df_clean[col].unique())


=== Constant Columns (no variance) ===
['Z_CostContact', 'Z_Revenue']

✅ Original columns: 29
✅ Columns removed:  2
✅ Columns remaining: 27
Z_CostContact: unique values -> [3]
Z_Revenue: unique values -> [11]


## 4. Add a new column (15pts)

* Add a new column **Customer_Days** to store the total number of days each customer has been with the company. You may follow the steps below:
  * use `pd.to_datetime(df["Dt_Customer"])` to convert the string variable to a *date-time* variable

  * use `today = datetime.today()` to find the date of today

  * use `(today - df["Dt_Customer"]).dt.days` to find the difference between today and the onbard date.

## 5. Categorical Features (15 pts)

* For `Education` and `Marital_Status`,

  * find how many customers fall into each category
  * create pie charts to show the percentage of each category; make sure to show the percentage numbers on the chart with 2 decimal places

* What do you findings suggest about the company's target demographic?

## 6. Numeric Variables (15pts)

* For `Income`, `Kidhome`, and `Teenhome`,
  * use boxplots to visualize their distributions.

* Based on these distributions, what can you say about the typical household profile of Nata's customers?

## 7. Spending by Products (15pts)

* Examine the spending amount columns `MntWines`,	`MntFruits`,	`MntMeatProducts`,	`MntFishProducts`,	`MntSweetProducts`,	`MntGoldProds`, and find the product with the highest total spending amount by all customers.



## 8. Use of Gen. AI (2pts)

Use of Generative AI is allowed. Briefly describe if and how you have used Gen. AI tools in this assignment.